File: Examples/test-suite/perl5/argout_runme.pl
use strict;
use warnings;
use Test::More tests => 9;
BEGIN { use_ok('argout') }
require_ok('argout');

sub check { my($c, $l) = @_;
  my $ptr = argout::new_intp();
  argout::intp_assign($ptr, 5);
  is($c->($ptr), 5, "rv $l");
  is(argout::intp_value($ptr), 6, "av $l");
  argout::delete_intp($ptr);
}

check(\&argout::incp, 'incp');
check(\&argout::incr, 'incp');
check(\&argout::inctr, 'incp');

# TODO:
# if we add this to argout.i
#    %inline %{
#      typedef void *vptr;
#    %}
#    %pointer_functions(vptr, vpp);
# then this test will work
SKIP: {
  skip 'need voidpp handler', 1, unless argout->can('new_vpp');
  my $vpp = argout::new_vpp();
  argout::voidhandle($vpp);
  my $vp = argout::vpp_value($vpp);
  argout::delete_vpp($vpp);
  is(argout::handle($vp), 'Here it is', 'opaque void ptr');
}

File: Examples/test-suite/lua/argout_runme.lua
local v=require("argout")

function doTest(incFunc)
  -- t = new int
  local t = v.new_intp()
  -- (*t) = 5
  v.intp_assign(t, 5)
  -- return (*t)++
  local val1 = incFunc(t)
  assert(val1 == 5)
  -- return *t
  local val2 = v.intp_value(t)
  assert(val2 == 6)
end

doTest(v.incp)  -- increase using pointer
doTest(v.incr)  -- increase using reference
doTest(v.inctr) -- increase using type defined reference
